### ISSUE TITLE
Provide a legitimate save/load mechanism

### ISSUE THREAD

**Issue by phreed** [webgme/webgme#895]
The current export/import is lossy.
For example:
given 
 * three objects which inherit as shown T1 <- T2 <- T3 
 * T1 contains two document nodes D1A and D1B which contain content.
 * T2 does not modify its copies (D2A and D2B) of those nodes.
 * T3 modifies its copies (D3A and D3B).

The export does include D1A, D1B, D3A and D3B, but not D2A nor D2B.
The serialization of D3* refereces a "base" guid that is not present in the export.
When the file is imported the proper relationship between the nodes is lost and some patching is done.
Sometimes the patch is correct and sometimes it is not.

Bottom line there is no guarantee that the model you import is the same as the export.
The export file *is-not* an effective way to save models.

**Comment by pmeijer** [comment]
We are planning to move away from the current export format and export the raw data instead. That task is targeted for 1.7.0. In the meantime the safest way to backup the data is to do a mongodump.

**Comment by kecso** [comment]
Do you have any examples when you lost any data because of the current export format?
However it is true that there is no explicit document nodes, it is not because we lose any information!

**Comment by phreed** [comment]
I think I have seen the case described above effectively swapping the bases of D3A and D3B.
Based on your description of how it works (and my understanding) this defect would be non-deterministic.

**Comment by phreed** [comment]
How do I do a mongodump?

**Comment by pmeijer** [comment]
https://docs.mongodb.org/manual/tutorial/backup-and-restore-tools/

**Comment by kecso** [comment]
The current export/import is deterministic. If you experience it otherwise please provide an example and we will fix it. (the upcoming changes to the export/import in this case do not matter as it can mean that current projects are in danger...)
The scenario you mentioned (swapping bases) is highly unlikely, if it would occur, then a simple 're-export' would point that out as the created file content would change though there were no change in the project(just letting you know how you can check it if you would suspect such error)...

**Comment by pmeijer** [comment]
Fred when you have tried the new export format (introduced as part of the libraries #925) can you get back to this issue and determine if it can be closed or not?

**Comment by phreed** [comment]
I made an example to demonstrate this problem (see attached).
But I do not know how to read the webgmex/project.json file.
[orphan_objects.json.txt](https://github.com/webgme/webgme/files/224217/orphan_objects.json.txt)

**Comment by kecso** [comment]
The current export format is not intended to use as an exchange for other tools. It is more of an internal serialization for WebGME only (you should not look into the webgmex or use it other than importing to another project or seeding or as a library).
Right now, if you need an exchange format, you should create a plugin that traverses your project and creates the output of your needs.
We are planning to implement a generic XMI format export, but the schedule is not decided yet.

Please let us know if that proposed feature would fulfill your needs!

**Comment by phreed** [comment]
The export format provides a mechanism for preserving work. I want to be sure that it can do its job. When I see things like a guid in the web visualization but when I search the webgmex/project.json file and do not find it, it bothers me.

**Comment by kecso** [comment]
Right now, the project package contains the raw data from the database that is not 'readable' as many part of the actual model interpretation is computed on the fly (like the guid for example).
To check that no data was lost, you should use the file as an import and decide on the basis of the resulting project.

Please let us know if you find an example of lost data!

**Comment by phreed** [comment]
I expect the guids to remain across an export/import. 
The guids are an important part of my model.
My export writes out the guids are you telling me they are not reliable?

**Comment by kecso** [comment]
No. I was telling that you should not try to look for any information directly in the export file (created by WebGME).
Your guids should remain intact during the export/import process as well as any other information your project holds...
